scope_shared/scope/populate/data/validate.py

```diff
@@ -0,0 +1,424 @@
+            }
+        ).unique()
+
+        datetime_scheduled_activity = datetime_from_document(document=document_current)
+        datetime_activity_schedule = datetime_from_document(document=activity_schedule_current)
+        timedelta_difference = (datetime_scheduled_activity - datetime_activity_schedule).total_seconds()
+
+        # 5 seconds would be really slow, most of these are 0 or 1
+        assert timedelta_difference < 5
+
+    for document_current in scheduled_activity_documents.filter_match(
+        match_deleted=False,
+    ):
+        # A referenced activity schedule might no longer exist.
+        # This is because we do not update scheduled activities in the past.
+        # A referenced activity schedule must have existed at some point.
+        # Its snapshot must be current or what it was before deletion.
+        activity_schedule_snapshot = activity_schedule_documents.filter_match(
+            match_datetime_at=datetime_from_document(document=document_current),
+            match_values={
+                "_set_id": document_current["activityScheduleId"]
+            }
+        ).unique()
+        if activity_schedule_snapshot.get("_deleted"):
+            # Snapshot should be of the state immediately before deletion.
+            assert activity_schedule_snapshot["_deleted"]
+
+            activity_schedule_snapshot = activity_schedule_documents.filter_match(
+                match_deleted=False,
+                match_values={
+                    "_rev": activity_schedule_snapshot["_rev"] - 1,
+                    "activityScheduleId": document_current["activityScheduleId"]
+                }
+            ).unique()
+
+        assert document_current["dataSnapshot"]["activitySchedule"] == activity_schedule_snapshot
+
+        # The referenced activity must exist.
+        # The snapshot of the activity must match.
+        activity_snapshot = activity_documents.filter_match(
+            match_datetime_at=datetime_from_document(document=document_current),
+            match_values={
+                "_set_id": activity_schedule_snapshot["activityId"]
+            }
+        ).unique()
+        if activity_snapshot.get("_deleted"):
+            # Snapshot should be of the state immediately before deletion.
+            assert activity_snapshot["_deleted"]
+
+            activity_snapshot = activity_documents.filter_match(
+                match_deleted=False,
+                match_values={
+                    "_rev": activity_snapshot["_rev"] - 1,
+                    "activityId": activity_schedule_snapshot["activityId"],
+                }
+            ).unique()
+
+        assert document_current["dataSnapshot"]["activity"] == activity_snapshot
+
+        # If the activity has a value, the referenced value must exist.
+        # If the activity has a value, the snapshot of the activity must match.
+        if "valueId" in activity_snapshot:
```

Comment:
Do we need to ensure document_current["dataSnapshot"]["value"] does not exist `if "valueId" is not in activity_snapshot`?

Reply:
Sure, added in c98fb8f4e339c0db4db7190d29b4a51d4c503592